lib/active_model/serializer.rb

```diff
@@ -188,7 +188,9 @@ def as_json(adapter_opts = nil)
 
     # Used by adapter as resource root.
     def json_key
-      root || _type || object.class.model_name.to_s.underscore
+      root ||
+        (_type && _type.is_a?(Proc) ? _type.call(object) : _type) ||
+        object.class.model_name.to_s.underscore
     end
```

Comment:
let's not complicate the code.  You can already pass in `root` or override `json_key` or `_type`. Better just to document this (esp until we do the clean up code for this... i.e. make it simpler!)